### PR TITLE
Adjust DataTable Header/Footer on Body overflow

### PR DESCRIPTION
**Pull request by leossantiago**
#### What does this PR do?

This fix forces `Header` and `Footer` of `DataTable` to adjust themselves according to `Body` content overflow (and showing a scrollbar).

By setting a Ref to Body's content, we can track its width and its parent's width. This might be important when dealing with custom scrollbars. The difference in size, saved as a state, is passed down to Header and Footer, where it's used inside their widths calculations.

#### Where should the reviewer start?

`src/js/components/DataTable/DataTable.js`

#### What testing has been done on this PR?

Mainly storybook using its examples.

#### How should this be manually tested?

Use any DataTable example (i.e. https://storybook.grommet.io/?path=/story/visualizations-datatable--clickable-data-table) and pass `size="small"` (or any size which leads to an overflow, considering the length of the provided data) as prop to DataTable.

#### What are the relevant issues?

Issue #4887

#### Screenshots (if appropriate)
Before:
image

After:
image

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible functionality-wise. Breaking change considering Header and Footer width shrink.

### CODE REVIEW

File: src/js/components/DataTable/DataTable.js [hunk 1]
```diff
@@ -127,10 +127,22 @@ const DataTable = ({
 
   // placeholder placement stuff
   const headerRef = useRef();
+  const bodyRef = useRef();
   const footerRef = useRef();
   const [headerHeight, setHeaderHeight] = useState();
   const [footerHeight, setFooterHeight] = useState();
 
+  const [overflowComp, setOverflowComp] = useState(0);
```
Comment (ShimiSun) [inline src/js/components/DataTable/DataTable.js:135]: would it make sense to call this state `bodyOverflow` vs `overflowComp`?

**Review by ShimiSun — COMMENTED** [review]
Overall looks good to me and that it is doing the trick. Dropped a small comment regarding readability, let me know what you think.
💟

File: src/js/components/DataTable/DataTable.js (hunk 1 above)
Comment (leossantiago) [reply src/js/components/DataTable/DataTable.js:135]: I was going for `overflowCompensation` but I do agree it could use some work on readability. I like `bodyOverflow` but I'm not sure if it's related to what it does. Even though `body` is the one overflowing, I don't think it would explain its use.

WDYT about `scrollBarWidth` or maybe `scrollBarCompensation` ?

Also, I'll try pushing at least one snapshot later on, so we can keep track of this behavior.

Comment (ericsoderberghp) [reply src/js/components/DataTable/DataTable.js:135]: We've used the term `offset` for situations like this elsewhere. Perhaps `scrollOffset`?

**Comment by leossantiago** [comment]
While adding tests to this, I noticed GroupedBody needed a fix as well, but since its overflow can be triggered after rendering and wouldn't notify us about changes to the current ref value, the useEffect approach wouldn't solve it. So I migrated to a useCallback ref. That way we can use the same code to both Body and GroupedBody. 

I'd appreciate your opinions on that, @ShimiSun @ericsoderberghp ! 

Body:
image

GroupBody:
Peek 2021-01-25 23-32

**Comment by leossantiago** [comment]
**NOTE:** removed tests. It seems jest/RTL make use of jsdom, which can't handle what this code does. It passes `width: calc(100% - 0px)` even on examples that should overflow, making those tests useless IMHO. Let me know what you think about that as well, and if there's another approach we could try to test this. Otherwise, we'll probably stick to manual testing done with storybook.

File: src/js/components/DataTable/DataTable.js [hunk 2]
```diff
@@ -131,6 +131,16 @@ const DataTable = ({
   const [headerHeight, setHeaderHeight] = useState();
   const [footerHeight, setFooterHeight] = useState();
 
+  // offset compensation when body overflows
+  const [scrollOffset, setScrollOffset] = useState(0);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const bodyRef = useCallback(ref => {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/DataTable.js:138]: One suggestion, can we do something like this instead?
```
  const bodyRef = useRef();
  useLayoutEffect(() => {
    const nextScrollOffset =
      bodyRef.current.parentElement?.clientWidth - bodyRef.current.clientWidth;
    if (nextScrollOffset !== scrollOffset) setScrollOffset(nextScrollOffset);
  });

Comment (leossantiago) [reply src/js/components/DataTable/DataTable.js:138]: Sure thing! I'll push a commit addressing that! Also, thanks for pointing that out. Never had a use case for useLayoutEffect so I've been reading about it a little bit 😄

**Review by ericsoderberghp — APPROVED** [review]

**Comment by ericsoderberghp** [comment]
Thanks!

**Comment by ShimiSun** [comment]
Thank you @leossantiago 💟
